pkg/controller/webhookbootstrap/controller.go

```diff
@@ -0,0 +1,488 @@
+	if err != nil {
+		log.Error(err, "Error decoding CA certificate data")
+		return err
+	}
+
+	// read the existing private key
+	pkData := readSecretDataKey(secret, corev1.TLSPrivateKeyKey)
+	if pkData == nil {
+		log.Info("Generating new private key")
+		return c.generatePrivateKey(crt, secret)
+	}
+	pk, err := pki.DecodePrivateKeyBytes(pkData)
+	if err != nil {
+		log.Info("Regenerating new private key")
+		return c.generatePrivateKey(crt, secret)
+	}
+	// read the existing certificate
+	if !c.certificateRequiresIssuance(ctx, log, secret, pk, crt) {
+		log.Info("Serving certificate already up to date")
+		return nil
+	}
+
+	// TODO: check to make sure the serving certificate is signed by the CA
+	certData, err := c.signCertificate(crt, pk, caPK, caCert)
+	if err != nil {
+		log.Error(err, "error signing certificate")
+		return err
+	}
+
+	return c.updateSecret(secret, pkData, caCertData, certData)
+}
+
+func (c *controller) certificateRequiresIssuance(ctx context.Context, log logr.Logger, secret *corev1.Secret, pk crypto.Signer, crt *cmapi.Certificate) bool {
+	// read the existing certificate
+	crtData := readSecretDataKey(secret, corev1.TLSCertKey)
+	if crtData == nil {
+		log.Info("Issuing webhook certificate")
+		return true
+	}
+	cert, err := pki.DecodeX509CertificateBytes(crtData)
+	if err != nil {
+		log.Info("Re-issuing webhook certificate")
+		return true
+	}
+
+	// ensure private key is valid for certificate
+	matches, err := pki.PublicKeyMatchesCertificate(pk.Public(), cert)
+	if err != nil {
+		log.Error(err, "internal error checking certificate, re-issuing certificate")
+		return true
+	}
+	if !matches {
+		log.Info("certificate does not match private key, re-issuing")
+		return true
+	}
+
+	// validate the common name is correct
+	expectedCN := pki.CommonNameForCertificate(crt)
+	if expectedCN != cert.Subject.CommonName {
+		log.Info("certificate common name is not as expected, re-issuing")
+		return true
+	}
+
+	// validate the dns names are correct
+	expectedDNSNames := pki.DNSNamesForCertificate(crt)
+	if !util.EqualUnsorted(cert.DNSNames, expectedDNSNames) {
+		log.Info("certificate dns names are not as expected, re-issuing")
+		return true
+	}
+
+	// validate the ip addresses are correct
+	if !util.EqualUnsorted(pki.IPAddressesToString(cert.IPAddresses), crt.Spec.IPAddresses) {
+		log.Info("certificate ip addresses are not as expected, re-issuing")
+		return true
+	}
+
+	if c.certificateNeedsRenew(ctx, cert, crt) {
+		log.Info("certificate requires renewal, re-issuing")
+		return true
+	}
+
+	return false
+}
+
+func readSecretDataKey(secret *corev1.Secret, key string) []byte {
+	if secret.Data == nil {
+		return nil
+	}
+	d, ok := secret.Data[key]
+	if !ok {
+		return nil
+	}
+	return d
+}
+
+func (c *controller) generatePrivateKey(crt *cmapi.Certificate, secret *corev1.Secret) error {
+	pk, err := c.generatePrivateKeyBytes(crt)
+	if err != nil {
+		return err
+	}
+
+	return c.updateSecret(secret, pk, nil, nil)
```

Comment:
Do we want to update yet? There will typically be multiple updates in quick succession currently

Reply:
We return from the sync once writing the private key and generate the certificate in the second iteration currently, so yes.